Scroll indicators update - positioning correction, support for both indicators at the same time

janek opened the pull request:
Corrects positioning, adds configurable indicator insets, adds conditional positioning when one/both indicators are visible.

**Type of change:** Bug fix + feature

## Todos
* [x] corrected positioning
* [x] added support for both indicators at the same time 
* [x] implemented default/base apple inset values
* [x] added support for configurable `scrollIndicatorInsets` (as per apple docs)
* [x] fixed other subviews of UIScrollView being added in front of indicators
 
## Expected behavior
Scroll indicators should now position and style themselves in a way indistinguishable from Apple's. 

#### Testing Details
<!--- Please describe how you tested your changes. Remove if N/A -->
Tested visually on NativePlayer with some additional marker views for better confidence.
Tested with a smaller dedicated app, both visually and numerically.

### Screenshots 
Current (desired) positioning of indicators:

<img width="637" alt="screenshot 2019-03-05 at 13 58 14" src="https://user-images.githubusercontent.com/2146927/53807134-15aaeb80-3f4f-11e9-8f87-dcd5d17810be.png">




### Please check if the PR fulfills these requirements

* [x] Self-review: I am confident this is the simplest and clearest way to achieve the expected behaviour
* [x] There are no dependencies on other PRs or I have linked dependencies through Zenhub
* [ ] The commit messages are clean and understandable
* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Code runs on all relevant platforms (if major change: screenshots attached)

janek committed:
corrected positioning code

janek committed:
corrected positioning code

janek committed:
corrected positioning code

janek committed:
corrected positioning code

janek committed:
added a (hopefully temporary) decay term to offset calculation

janek committed:
added a (hopefully temporary) decay term to offset calculation

janek committed:
added a (hopefully temporary) decay term to offset calculation

janek committed:
added a (hopefully temporary) decay term to offset calculation

janek committed:
included additional insets in the positioning, eliminated decayTerm

janek committed:
included additional insets in the positioning, eliminated decayTerm

janek committed:
included additional insets in the positioning, eliminated decayTerm

janek committed:
included additional insets in the positioning, eliminated decayTerm

janek committed:
ensured default insets load appropriately, added small fixes

janek committed:
ensured default insets load appropriately, added small fixes

janek committed:
ensured default insets load appropriately, added small fixes

janek committed:
ensured default insets load appropriately, added small fixes

janek committed:
lay out indicators out to their initial position when new insets are set from code

janek committed:
lay out indicators out to their initial position when new insets are set from code

janek committed:
lay out indicators out to their initial position when new insets are set from code

janek committed:
lay out indicators out to their initial position when new insets are set from code

janek committed:
Corrected indicators positioning

janek committed:
Corrected indicators positioning

janek committed:
Corrected indicators positioning

janek committed:
Corrected indicators positioning

janek committed:
Added initial distance from scrollview frame

janek committed:
Added initial distance from scrollview frame

janek committed:
Added initial distance from scrollview frame

janek committed:
Added initial distance from scrollview frame

janek committed:
support for both indicators at once

janek committed:
support for both indicators at once

janek committed:
support for both indicators at once

janek committed:
support for both indicators at once

janek committed:
additional spacing that prevents overlap when both indicators are present

janek committed:
additional spacing that prevents overlap when both indicators are present

janek committed:
additional spacing that prevents overlap when both indicators are present

janek committed:
additional spacing that prevents overlap when both indicators are present

janek committed:
connected confiugrable insets code to placement code

janek committed:
connected confiugrable insets code to placement code

janek committed:
connected confiugrable insets code to placement code

janek committed:
connected confiugrable insets code to placement code

janek committed:
abstracted out 1 additional step for clarity, corrected var privacy

janek committed:
abstracted out 1 additional step for clarity, corrected var privacy

janek committed:
abstracted out 1 additional step for clarity, corrected var privacy

janek committed:
abstracted out 1 additional step for clarity, corrected var privacy

janek committed:
correction for overlap-avoidance, style improvements

janek committed:
correction for overlap-avoidance, style improvements

janek committed:
correction for overlap-avoidance, style improvements

janek committed:
correction for overlap-avoidance, style improvements

janek committed:
corrected syntax bug

janek committed:
corrected syntax bug

janek committed:
corrected syntax bug

janek committed:
corrected syntax bug

janek committed:
Merge branch 'scroll-indicators-update' of github.com:flowkey/UIKit-SDL into scroll-indicators-update

janek committed:
ensured scroll indicators stay on top of all other subviews of UIScrollView

janek committed:
ensured scroll indicators stay on top of all other subviews of UIScrollView

ephemer committed:
Merge branch 'master' into scroll-indicators-update

ephemer committed:
Implement some PR feedback

ephemer committed:
Implement some PR feedback

janekszynal commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -12,26 +12,55 @@
-            horizontal: contentInset.left + contentSize.width + contentInset.right,
-            vertical: contentInset.top + contentSize.height + contentInset.bottom
-        )
+    internal var indicatorLengths: (horizontal: CGFloat, vertical: CGFloat) {
```
clear about get{}, just slipped my mind that it's legal without it - correcting in the next commit.

controlled access levels and I was able to change a few to private, leaving only three functions that are indeed called from outside the file (from `UIScrollView.swift` in case of all 3)

janek replied:
Access levels were still not corrected, must have not pushed some changes. pushed them now.

janekszynal commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -12,26 +12,55 @@
-        )
+    internal var indicatorOffsetsInContentSpace: (horizontal: CGFloat, vertical: CGFloat) {
+        get {
+            let indicatorDistanceFromScrollViewFrame =  UIScrollView.indicatorDistanceFromScrollViewFrame
```
there was a conscious effort there (to have all numerical constants at the top of the file) - happy to discuss sometime, but for now just changing to what you suggested

janekszynal commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -12,26 +12,55 @@
+                vertical: (contentInset.top + contentOffset.y) / (totalContentArea.vertical - bounds.height)
+            )
+
+            let bothIndicatorsShowing = shouldLayoutHorizontalScrollIndicator && shouldLayoutVerticalScrollIndicator
```
renamed.
think I assigned the right meaning to how I was thinking about that constant, but maybe had a wrong picture of that

janekszynal commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -12,26 +12,55 @@
+            let bothIndicatorsShowing = shouldLayoutHorizontalScrollIndicator && shouldLayoutVerticalScrollIndicator
+
+            let additionalSpacingToPreventOverlap = (
+                horizontal: bothIndicatorsShowing ? 2*indicatorDistanceFromScrollViewFrame : 0,
```
done

janek committed:
Merge branch 'scroll-indicators-update' of github.com:flowkey/UIKit-SDL into scroll-indicators-update

janekszynal commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -42,28 +71,27 @@ internal extension UIScrollView {
-                x: indicatorDistanceFromScrollViewFrame + indicatorOffsets.horizontal,
-                y: bounds.height - (indicatorThickness + indicatorDistanceFromScrollViewFrame),
+                x: scrollIndicatorInsets.left + indicatorOffsetsInContentSpace.horizontal, //scrolling direction
+                y: bounds.size.height + contentOffset.y - distanceFromFrame.horizontal,
```
done

janekszynal commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -42,28 +71,27 @@ internal extension UIScrollView {
             horizontalScrollIndicator.frame = CGRect(
-                x: indicatorDistanceFromScrollViewFrame + indicatorOffsets.horizontal,
-                y: bounds.height - (indicatorThickness + indicatorDistanceFromScrollViewFrame),
+                x: scrollIndicatorInsets.left + indicatorOffsetsInContentSpace.horizontal, //scrolling direction
```
this was just for me and should have been removed before pushing. sorry, fixed.

janek committed:
more PR feedback changes

janek committed:
more PR feedback changes

janekszynal commented on Sources/UIScrollView.swift:
```diff
@@ -120,10 +130,29 @@ open class UIScrollView: UIView {
+    
+    public var scrollIndicatorInsets: UIEdgeInsets {
+        get {
+            let totalScrollIndicatorInsets = UIEdgeInsets(top: baseScrollIndicatorInsets.top + customScrollIndicatorInsets.top,
```
you're right, this needs correction, I'm working on it now. 

They are indeed 0,0,0,0 when logged. 
I took my values from observing the placement, they are indeed inset 2.5 from the sides - but when it's set to 0. It's a bit confusing, but I'll come up with something.

janekszynal replied:
ok, done, this actually made code clearer aswell. thanks for noticing it!

janek committed:
corrected UIEdgeInsets code

janek committed:
corrected UIEdgeInsets code

ephemer commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -42,28 +67,27 @@ internal extension UIScrollView {
-        let indicatorOffsets = indicatorOffsetsInContentSpace()
-        let indicatorLengths = (horizontal: (bounds.width / contentSize.width) * bounds.width,
-                                vertical: (bounds.height / contentSize.height) * bounds.height)
+        
```
whitespace

janek replied:
seems to be corrected/not exisiting. resolving.

janek committed:
Merge branch 'scroll-indicators-update' of https://github.com/flowkey/UIKit-cross-platform into scroll-indicators-update

janek committed:
rename 'bothIndicatorsShowing'

janek committed:
removed whitespace

rikner commented:
@ephemer @janekszynal is there anything preventing us from merging this? do we need any tests?

rikner committed:
Merge branch 'master' into scroll-indicators-update

rikner commented on Sources/UIScrollView.swift:
```diff
@@ -120,10 +127,24 @@ open class UIScrollView: UIView {
 
     let indicatorThickness: CGFloat = 2.5
+    
+    public var indicatorStyle: UIScrollViewIndicatorStyle = .`default` {
```
`UIScrollViewIndicatorStyle` is now `UIScrollView.IndicatorStyle`: https://developer.apple.com/documentation/uikit/uiscrollview/indicatorstyle

janek replied:
done

rikner reviewed (commented):
@janek maybe you can add a screenshot of your testapp, which shows both indicators. apart from that and the little API change we need to conform to I think this is ready to get merged soon.

ephemer commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -42,28 +67,27 @@ internal extension UIScrollView {
 
-
     internal func layoutScrollIndicatorsIfNeeded() {
         guard shouldLayoutHorizontalScrollIndicator || shouldLayoutVerticalScrollIndicator else { return }
```
iOS always lays out the scroll indicators I think, and there was an advantage to this -> I seem to remember finding a bug where the indicators appeared in the wrong spot after being hidden. I think we need to revisit this

janek replied:
hey, I ran a test app with a scrollview on pure iOS, with showsXScrollIndicator off for both dimensions, and then on. I then used `po scrollView.subviews` in the debugger. If showsXScrollIndicator is off, they indicators are not in the view hierarchy at all, so it seems like iOS does not position them in that case. I could imagine that iOS might save positions for indicators without adding them to view hierarchy just yet, so maybe that's what happens - but no idea how to test that.

In light of this, what do we want to do? 
From the perspective of code changes, laying them out in all cases should be very easy and help us get rid of a few lines of code. So I absolutely don't mind doing it if you feel that's the way to go.

ephemer replied:
Hi Janek, you can get a reference to the scroll indicator subviews (just iterate through them on an empty scroll view and find the UIImageViews)  and then hide them. You should be able to print their positions at every pixel scrolled. Would be cool if you could check this. I think it makes sense to remove them from the hierarchy of they’re hidden, but laying them out shouldn’t be a big performance issue (if that’s what iOS does)

janek replied:
yes, I can get the reference and print positions, even if I make them invisible with `isHidden`.

but if I set `showsVerticalScrollIndicator` for `false`, that indicator will no longer be in the view hierarchy and therefore probably impossible to get a reference to.

I checked back with our code and this is where we differ - we always have indicators in the hierarchy and just make them invisible if the UIKit user wants to hide them.

So if they are set to hidden by UIKit API, we keep them in the hierarchy, but don't update their positions, and then start updating it when they become visible. We could:
1. keep it as it is 
2. always update their positions, even when theyr are invisible
3. try to get as close to iOS as possible and actually remove them and add back as needed.

1. and 2. are equally easy. We are in state 1 and if we want 2, we should just merge this new small PR I opened (#284). It's set to merge into this branch, not master.

3. is more work and I worry it might also complicate things with our inserting/removing subviews code, so I'd rather not do it now - but again, it seems the closest to iOS. Maybe let's go with 1 or 2 and create a low-prio issue for 3?

janek replied:
update: I now changed the insert/add subview code - didn't take long at all and should now be robust w.r.t indicators sometimes not being there. It therefore also shouldn't be too much pain to implement v3. I'll still wait for your opinion before starting on that.

janek replied:
update: merged #284

janek committed:
changed internal access levels to private in indicators extension, removed 'internal' keyword where unneccessarily explicit

janek committed:
updated UIScrollViewIndicatorStyle to be UIScrollView.IndicatorStyle (matching Apple's API change)

janek committed:
Added comments explaining source of thickness and inset values

janek commented on Sources/UIScrollView.swift:
```diff
@@ -120,10 +127,24 @@ open class UIScrollView: UIView {
+        didSet { applyScrollIndicatorsStyle() }
+    }
+    
+    private let baseScrollIndicatorInsets = UIEdgeInsets(top: 2.5, left: 2.5, bottom: 2.5, right: 2.5)
```
done

janek commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,22 @@ open class UIScrollView: UIView {
+    }
+    
+    open override func insertSubview(_ view: UIView, at index: Int) {
+        if index > subviews.count - 2 {
```
checked on it now and it's changed to your suggestion. resolving.

janek commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,19 @@ open class UIScrollView: UIView {
+        if view == horizontalScrollIndicator || view == verticalScrollIndicator {
+            super.addSubview(view)
+        } else {
+            super.insertSubview(view, at: subviews.count - 2)
```
I understand that this can be concerning. Also don't have anything breakthrough to propose, some small ideas: 
- make indicators their own class so it's more elegant (and safe?)  to check whether something is an indicator or not. 
- insted of inserting at end-2, we could check every element, starting from the last one, for being a scrollIndicator. if it is, we move one back. when it finally isn't, we insert. this should be a little safer.

ephemer replied:
I think we need to avoid hard coding 2 because if as you say the subviews are removed on iOS when hidden we might as well do that

What about the 
insertSubview(below: otherSubview) API?

ephemer commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -34,44 +59,43 @@ internal extension UIScrollView {
 
-
-    internal func layoutScrollIndicatorsIfNeeded() {
+    func layoutScrollIndicatorsIfNeeded() {
```
We have a convention of always writing `internal` to make it clear that that’s our intention. It’s the default in Swift but since we’re making a public-facing library our default is public/open.

janek replied:
Yeah, I was wondering why the `internals` are explicitly there. Guess I was on the wrong side of https://en.wikipedia.org/wiki/Wikipedia:Chesterton%27s_fence :d 

brought them back now

janek committed:
brought back explicit 'internal' access level keywords

janek committed:
removed conditions for laying out indicators, it will now happen every time. Renamed variables to fit this changed state logically.

janek committed:
improved safety of code responsible for adding subviews: we now check for indicators and always insert below them, instead of at a hardcoded index

janek commented:
hey @ephemer, I can't comment on your last comment, not sure why. It might be because I changed the code it was referencing (and it got auto-marked as outdated) or because it was a comment to the PR in general and not a line in code (and therefore doesn't have a reply thread).

Anyway, I changed the add/insert subview code. Couldn't avoid writing a few extra lines, but I think it's all pretty understandable and therefore no big harm done. Let me know what you think.

ephemer commented:
Hi @janek, I'd really like to get this PR merged because it's been open for a long time. My main concern is still this ordering of the indicators. I have the feeling that part could be buggy still. Let's take some time when you're back in the office to go through this and get it merged

janek commented:
yeah, I would love that too, it's been very long. Let's try tomorrow!

ephemer commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,50 @@ open class UIScrollView: UIView {
+    // Subview list goes back-to-front, so indicators need to take the last two positions.
+    // We use the earliestIndicatorInSubviewHierarchy variable to make sure new views are always inserted below them.
+    private lazy var earliestIndicatorInSubviewHierarchy: UIView? = {
+        let indexOfHorizontal = subviews.index(of: horizontalScrollIndicator)
```
this entire function can be written much more simply:

`return subviews.first(where: { $0 === horizontalScrollIndicator || $0 === verticalScrollIndicator })`

janek replied:
_nice_

ephemer commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,50 @@ open class UIScrollView: UIView {
+        if let earliestIndicatorInSubviewHierarchy = earliestIndicatorInSubviewHierarchy {
+            indexOfLastIndicator = subviews.index(of: earliestIndicatorInSubviewHierarchy)!
+        }
+        super.insertSubview(view, at: min(index, indexOfLastIndicator-1))
```
linting: `indexOfLastIndicator - 1`

janek replied:
corrected

ephemer commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,50 @@ open class UIScrollView: UIView {
+    open override func addSubview(_ view: UIView) {
+        // If the view we're adding is an indicator, just add it.
+        // If not, check for indicators and insert below the last one present.
+        if view == horizontalScrollIndicator || view == verticalScrollIndicator {
```
this can be simplified too:

```
if view !== horizontalScrollIndicator, view !== verticalScrollIndicator, let earliestIndicatorInSubviewHierarchy = earliestIndicatorInSubviewHierarchy  {
    super.insertSubview(view, belowSubview: earliestIndicatorInSubviewHierarchy)
} else {
    super.addSubview(view)
}
```

janek replied:
done

ephemer commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,50 @@ open class UIScrollView: UIView {
+        if view == horizontalScrollIndicator || view == verticalScrollIndicator {
+            super.addSubview(view)
+        } else {
+            if let earliestIndicatorInSubviewHierarchy = earliestIndicatorInSubviewHierarchy {
```
I'm not sure on the naming of this: `subviewHierarchy` is overly descriptive to me, and `earliest` is imprecise. What does earliest even mean in terms of array indexes or visual hierarchy? Is it `firstIndicatorInSubviews` maybe?

janek replied:
you're right, `firstIndicatorInSubviews` is way better (especially being consistent with the `.first(where:{}`) function). changed

ephemer commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,50 @@ open class UIScrollView: UIView {
+
+
+    open override func insertSubview(_ view: UIView, at index: Int) {
+        var indexOfLastIndicator = Int.max
```
I'm not a big fan of this mutable style. I think it's harder to read as is, and prone to errors if we need to change it.

I think we can just have two code paths, one that does the `min` comparison, and one that just adds the view at the specified index.

janek replied:
done

janek replied:
I agree with this point and refactored - should be nice and readable now.

ephemer commented on Sources/UIScrollView.swift:
```diff
@@ -120,11 +158,27 @@ open class UIScrollView: UIView {
+    private let baseScrollIndicatorInsets = UIEdgeInsets(top: 2.5, left: 2.5, bottom: 2.5, right: 2.5)
+
+    public var indicatorStyle: IndicatorStyle = .`default` {
+        didSet { applyScrollIndicatorsStyle() }
```
just making sure `applyScrollIndicatorsStyle` is called elsewhere (as well) - `didSet`s don't get called on `init`. If this is the case we should add a comment to that effect here too

janek replied:
yes, it is being called on `init` separately. Not sure if I understand you correctly, should I add a comment stating that fact?

ephemer commented on Sources/UIScrollView+indicatorsInternal.swift:
```diff
@@ -6,26 +6,51 @@
+
+        let shouldShowBothIndicators = shouldLayoutHorizontalScrollIndicator && shouldLayoutVerticalScrollIndicator
+
+        let additionalSpacingToPreventOverlap = (
```
Did we make these distances up or are they iOS style? We should comment which

janek replied:
done

janek commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,50 @@ open class UIScrollView: UIView {
+    }()
+
+
+    open override func addSubview(_ view: UIView) {
```
I'm realizing that I should have overriden the `subviews` variable. Not only does it probably make `insertSubview` simpler, but should also just be there. Currently the behavior for `print(scrollView.subviews.count)` will differ from iOS in that it will also count the indicators.

Unfortunately, if I do override it, those versions of `addSubview` and `insertSubview` won't work as expected. I'm trying to figure it out now, noting here so all information about remaining todos is in one place.

janek replied:
Update: no need to override scrollviews, iOS does have them

ephemer replied:
@janek here you reported something different regarding subviews, or did I misunderstand one of the two pieces of info incorrectly :)

janek committed:
Test push

janek committed:
Removing testfile

janek commented on Sources/UIScrollView.swift:
```diff
@@ -77,6 +71,50 @@ open class UIScrollView: UIView {
+    }()
+
+
+    open override func addSubview(_ view: UIView) {
```
consider:
```
super.addSubview(view)
addHorizontal
addVertical
```
for much simpler, a bit less performant code

janek replied:
not sure what exactly I meant here, unfortunately - resolving thread

janek committed:
Merge branch 'master' into scroll-indicators-update

# Conflicts:
#	Sources/UIScrollView+indicatorsInternal.swift - resolved using this branch's version

janek committed:
Linting: spaces should be on both sides of the minus sign

janek committed:
Added a comment about distance values mimicking iOS

janek committed:
Temporary changes to the Demo App

janek committed:
Add a test for scroll veiw with scroll indicators subview hierarchy

janek committed:
Added information about indicators being accesible in `subviews` and a commented-out stub that could help fix that later

janek committed:
Fix mockTouch

janek committed:
Add a guard to 'layout indicators if necessary'

janek committed:
FIx whitespace

janek committed:
simplified `earliestIndicatorInSubviewHierarchy`

janek committed:
Improve subview count test

janek committed:
Remove the stub for overriding `subviews`, as default is correct

janek committed:
Update DemoApp

janek committed:
Improve style in addSubview

janek committed:
Add more hierarchy positions tests

janek committed:
Improve tests, refactor insertSubview

janek committed:
Add a comment stating that `applyScrollIndStyle` is called on init (and not only didSet)

janek committed:
Update the DemoApp to change black to blue

janek commented:
hey @ephemer :) I improved a lot here, addressing all the smaller comments you had. 

Unfortunately, I still have a concern:

if you check `scrollView.subviews` on iOS, it will not include scroll indicators. In our version, in the state from this PR, it will include them. We can address this here or with a new issue, depending on how demanding/important it is. I gave it a moment's thought and decided it's best to ask you for opinion before trying, and a tip on how to go about it.

ephemer commented:
@janek we could override `var subviews` to return a filtered version of `private var _subviews = [UIView]`. It's tricky though because AFAIK we're using `subviews` for some "internal" UIKit logic, which may or may not be correct.

But to me that's an issue that's totally separate from this PR. If you would like to update that let's do it for a practical (rather than idealistic) reason: right now I don't think it matters for us

janek commented:
cool, that's what I was thinking, too, for the implementation. Glad I had the right idea (and the right idea to ask). 

In that case, this is ready for re-review.

ephemer commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,65 @@ class UIScrollViewTests: XCTestCase {
+        let indexOfInsertedView2 = scrollView.subviews.index(of: viewToBeInserted1)!
+
+        // If we try to insert at a position occupied by or above indicators,
+        // the inserted view should 'slide down' and assume the highest position below indicators
```
nice, these comments are helpful thanks!

ephemer replied:
but this should just be its own standalone test. it tests one single behaviour that obviously must be true. if we insert a subview it should always be behind the indicators. end of test. if that breaks we know what to fix. does that make sense?

ephemer commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,65 @@ class UIScrollViewTests: XCTestCase {
+        let indexOfInsertedView1 = scrollView.subviews.index(of: viewToBeInserted1)!
+
+        // This might seem weird, but that's what happens in iOS:
+        // the inserted view and the one that was in its place before will have the same index
```
is it possible this is only weird because of how the test is written?

It doesn't seem that weird to me that the indexes will overlap once we insert a new subview if we're not getting the _new_ indexes after inserting the subview?

What am I missing here? It seems like we should get the _new_ indexes here which would be less confusing?

ephemer commented on samples/getting-started/DemoApp/ViewController.swift:
```diff
@@ -15,12 +15,31 @@ class ViewController: UIViewController {
+        }
+
+
+        let blueView = UIView(frame: CGRect(origin: CGPoint(x: 225+20*2 - 50, y: 15+20*2), size: CGSize(width: 150, height: 10)))
```
please always use a space between binary operators
also, this line is very long, it would be great if you could split it up logically into

`let blueViewFrame = CGRect(x:,y:,width:,height:)` (avoiding the origin:,size: initializer)

also, these values seem pretty arbitrary, would be good if you could put them into named variables so they get some meaning

janek replied:
(pasting in both relevant places) 
Very sorry about this, the changes to this file were not meant to be submitted. They were useful to me while developing, but my goal wasn't to change the DemoApp. I was pushing it to keep the git index clean, with the intention of resetting the file before re-requesting review.

ephemer commented on samples/getting-started/DemoApp/ViewController.swift:
```diff
@@ -15,12 +15,31 @@ class ViewController: UIViewController {
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
 
-        label.text = "Hello World"
```
please don't delete what we already had, but feel free to move the label somewhere else (outside the scroll view)

janek replied:
(pasting in both relevant places) 
Very sorry about this, the changes to this file were not meant to be submitted. They were useful to me while developing, but my goal wasn't to change the DemoApp. I was pushing it to keep the git index clean, with the intention of resetting the file before re-requesting review.

ephemer replied:
@janek I think it's great to add more functionality to the demo app, totally fine to leave it here, but please make the change additive rather than replacing what was there :)

ephemer commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,65 @@ class UIScrollViewTests: XCTestCase {
+
+
+        let viewToBeInserted2 = UIView()
+        scrollView.insertSubview(viewToBeInserted2, at: 5) // position currently occupied by scroll indicator
```
this test is too long. this could be two or three separate tests which would be much clearer if one of them fails

janek replied:
Thanks for this and all the other very valid concerns. I rewrote the test(s) in a way that hopefully takes into account all of them.

ephemer commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,65 @@ class UIScrollViewTests: XCTestCase {
+        XCTAssertEqual(indexOfInsertedView1, 1)
+        XCTAssertEqual(label2Index, 1)
+        XCTAssertEqual(label3Index, 2)
+        XCTAssertEqual(horizontalIndicatorIndex, 3)
```
This is misleading at best: all of these indexes will be incorrect now, so what's the point in asserting what their _old_ value was? I'm really not happy with this, as written

ephemer commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,65 @@ class UIScrollViewTests: XCTestCase {
+
+        XCTAssert(label2Index > label1Index)
+
+        XCTAssertEqual(label1Index, 0)
```
what's the point in adding three subviews? we already have other tests that test the behaviour of `addSubview`. we're just writing more test code that we have to maintain.

how is this different to the case at the bottom?

ephemer commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,65 @@ class UIScrollViewTests: XCTestCase {
+        let label2Index = scrollView.subviews.index(of: view2)!
+        let label3Index = scrollView.subviews.index(of: view3)!
+
+        XCTAssert(horizontalIndicatorIndex > label1Index)
```
how is this different to the case at the bottom?

ephemer commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,65 @@ class UIScrollViewTests: XCTestCase {
+
+        let horizontalIndicatorIndex = scrollView.subviews.index(of: scrollView.horizontalScrollIndicator)!
+        let verticalIndicatorIndex = scrollView.subviews.index(of: scrollView.verticalScrollIndicator)!
+        let label1Index = scrollView.subviews.index(of: view1)!
```
I don't think it makes sense to put this into a variable, because (as mentioned below) it will be invalid as soon as we insert another view.

```
XCTAssertGreaterThan(
    scrollView.subviews.index(of: scrollView.horizontalScrollIndicator),
    scrollView.subviews.index(of: view1)
)
```

ephemer reviewed (changes requested):
Hi @janek, code-wise the library code seems fine but the tests are very shonky as is.

The tests contain false assumptions and hardcoded outdated values which don't mean anything. Sometimes we're calling an identical assertion twice (what's the point in calling `XCTAssertEqual` on two static values twice?), even though the second time the reality is that the underlying values have changed. It's not up to scratch as written.

Please try to rewrite the new tests in more finer-grained ones that test _one_ thing, that are clearly named so it's obvious what to fix if they break, that don't rely on hardcoded values if possible, and that test the absolute minimal amount possible while still ensuring behaviour. And please be careful to get the _new_ indexes after changing the order of an array.

ephemer commented on samples/getting-started/DemoApp/ViewController.swift:
```diff
@@ -15,12 +15,31 @@ class ViewController: UIViewController {
+        blueView.backgroundColor = .blue
+
+        scrollView.insertSubview(blueView, at: 6)
+        print(scrollView.subviews.index(of: blueView) ?? -1)
```
if you do decide to clean this up and keep the code, please remove any `print`s

janek committed:
Restore ViewController for master

janek committed:
Divide test into two, clarify purpose and remove bug

janek committed:
Improve comment and test names

janek commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,56 @@ class UIScrollViewTests: XCTestCase {
+
+        // Currently broken
+        // XCTAssertEqual(indexOfInsertedView, mockViews.count)
+    }
```
Very happy to update this when #309 is merged

janek replied:
(updated)

janek committed:
Merge commit '8f10d29ea93f93ca65633c6bd1f1b863d2461b5a' into scroll-indicators-update

janek committed:
Update insertSubview test, fix a bug in insertSubview

janek commented:
thanks @ephemer! I didn't reply to all of your comments, but I took them into account when rewriting tests. It's definitely a big improvement, and I'm quite happy. Also very happy to perfect them if there is anything else that you see could be improved :)

ephemer commented on UIKitTests/UIScrollViewTests.swift:
```diff
@@ -47,6 +47,42 @@ class UIScrollViewTests: XCTestCase {
+        scrollView.addSubview(mockView)
+
+        XCTAssertGreaterThan(
+            scrollView.subviews.index(of: scrollView.horizontalScrollIndicator)!,
```
These tests look seriously amazing now. Great work @janek!

janek replied:
that's too nice! thanks a lot!!

ephemer reviewed (approved):
Awesome! I am stoked with how great these new tests you wrote are looking now! Thanks for your work on this @janek
